File: src/udemy_tutorials_cpp/src/add_two_ints_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

// This is just to not to have to put the full line when needed. So, that we
// can directly use them (in this case) by doing _1 and _2
using std::placeholders::_1;
using std::placeholders::_2;

class AddTwoIntsServer : public rclcpp::Node
{
    public:
        AddTwoIntsServer() : Node("add_two_ints_server")
        {
            server_ = this->create_service<example_interfaces::srv::AddTwoInts>(
                "add_two_ints",  // Service name
                std::bind(&AddTwoIntsServer::callback_add_two_ints, this, _1, _2));
            RCLCPP_INFO(this->get_logger(), "Service server has been started.");
        }
    private:
        void callback_add_two_ints(const example_interfaces::srv::AddTwoInts::Request::SharedPtr request,
                                   const example_interfaces::srv::AddTwoInts::Response::SharedPtr response)
        {
            response->sum = request->a + request->b;
            RCLCPP_INFO(this->get_logger(), "AddTwoInts: %ld + %ld = %ld", request->a, request->b, response->sum);
        }

        rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr server_;
};


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsServer>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/udemy_tutorials_py/udemy_tutorials_py/add_two_ints_server.py
#!/usr/bin/env python3
"""ROS2 service server node example."""
import sys
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


class AddTwoIntsServerNode(Node):
    """ROS2 service server node."""
    def __init__(self):
        """Class constructor to set up the node."""
        super().__init__('add_two_ints_server')
        self.create_service(
            srv_type=AddTwoInts,  # service type
            srv_name='add_two_ints',  # service name
            callback=self.callback_add_two_ints  # callback function
        )
        self.logger.info('Add Two Ints Server has been started.')

    @property
    def logger(self):
        """Logger function to log messages."""
        return self.get_logger()

    def callback_add_two_ints(self, request, response):
        """Callback function to handle the service request."""
        response.sum = request.a + request.b
        self.logger.info(f'Request: a={request.a}, b={request.b}')
        self.logger.info(f'Response: {response.sum}')
        return response


def main(args=None):
    """Main function to set up the node."""
    rclpy.init(args=args)
    add_two_ints_server = AddTwoIntsServerNode()
    rclpy.spin(add_two_ints_server)
    rclpy.shutdown()


if __name__ == '__main__':
    sys.exit(main())
